src/io_uring.c

```diff
@@ -0,0 +1,92 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+static size_t io_uring_queue_len = 0;
+
+void initIOUring(void) {
+    if (server.io_uring_enabled) {
+        struct io_uring_params params;
+        struct io_uring *ring = zmalloc(sizeof(struct io_uring));
+        memset(&params, 0, sizeof(params));
+        /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+         * point to the shared memory containing the io_uring queues.
+         * On failure -errno is returned. */
+        int ret = io_uring_queue_init_params(IO_URING_DEPTH, ring, &params);
+        if (ret != 0) {
+            /* Warning if user enable the io_uring in config but system doesn't support yet. */
+            serverLog(LL_WARNING, "Failed to initialize io_uring: %s", strerror(-ret));
+            zfree(ring);
+            server.io_uring = NULL;
+        } else {
+            serverLog(LL_NOTICE, "io_uring enabled.");
+            server.io_uring = ring;
+        }
+    }
+}
+
+int canFsyncUsingIOUring(void) {
+    return server.aof_state == AOF_ON && server.aof_fsync == AOF_FSYNC_ALWAYS && server.io_uring_enabled &&
+           server.io_uring;
+}
+
+/* Submit fdatasync request to io_uring. */
+void ioUringPrepFsyncAndSubmit(int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(server.io_uring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
+    io_uring_queue_len++;
```

Comment:
This counter can then be associated with the specific io_uring instance
```suggestion
    ((struct *io_uring) io_uring)->io_uring_queue_len++;
```

Reply:
@PingXie `struct io_uring` is defined in `<liburing.h>`. We can't add fields to it.

Reply:
Thanks! I missed the part that `io_uring` is already a public struct